### PR TITLE
feat:beam search biz decode

### PR DESCRIPTION
**Pull request by zuozhen-ali**

### CODE REVIEW

**Comment by LLLLKKKK** [comment]
🤖 AI Code Review — PR #855

## PR 概述

**Title:** `feat:beam search biz decode`
**Author:** zuozhen-ali
**规模:** 9(GitHub) files, +108/-0

### 核心目标
新增 `BeamDedupLogitsProcessor`，在 beam search 的指定 decode step 上对 logits 做去重 mask（只保留每个 beam 的 argmax token，其余置 -inf），用于业务侧 beam search 去重场景。变更涉及 Python/C++ GenerateConfig、proto、RPC 传输、logits processor 工厂注册。

---

## 改动逻辑拆解

### GitHub 开源仓库变更（主要代码）

#### 1. 配置层：beam_dedup_idx 字段传播
- `rtp_llm/config/generate_config.py`：Python `GenerateConfig` 新增 `beam_dedup_idx: Optional[int] = None`
- `rtp_llm/cpp/engine_base/stream/GenerateConfig.h`：C++ `GenerateConfig` 新增 `std::optional<int> beam_dedup_idx`
- `rtp_llm/cpp/model_rpc/proto/model_rpc_service.proto`：`GenerateConfigPB` 新增 `google.protobuf.Int32Value beam_dedup_idx = 101`
- `rtp_llm/cpp/model_rpc/QueryConverter.cc`：`TRANS_OPTIONAL(beam_dedup_idx)` 反序列化
- `rtp_llm/cpp/model_rpc/model_rpc_client.py`：`trans_option(...)` 序列化

#### 2. 核心逻辑：BeamDedupLogitsProcessor
- `BeamDedupLogitsProcessor.h`/`.cc`（新文件）：继承 `BaseLogitsProcessor`，在 `current_step_ == beam_dedup_idx_` 时对 logits 做 argmax mask
- `LogitsProcessorFactory.cc`：注册新 processor

---

## Checklist 检查结果

### 通用原则

#### 软件工程原则
| 检查项 | 结果 |
|--------|------|
| SRP | ✅ 单一职责，独立 processor |
| OCP | ✅ 通过工厂模式扩展 |
| LSP | ❌ 子类未遵守基类契约（见 P0-1） |
| ISP | ✅ |
| DIP | ✅ |
| DRY | ✅ |
| KISS | ✅ |
| YAGNI | ✅ |

#### 架构审视
| 检查项 | 结果 |
|--------|------|
| 抽象边界 | ❌ 使用 torch::Tensor API 但基类接口是 BufferPtr（见 P0-1） |
| 依赖方向 | ✅ |
| 状态完整性 | ✅ current_step_ 单调递增，无需 reset |
| 错误语义 | ✅ 参数非法时 warning + 返回 nullptr |
| 可观测性 | ✅ 有 warning 日志 |
| 可演进性 | ✅ |
| 可运维性 | ✅ |

#### 测试
| 检查项 | 结果 |
|--------|------|
| 新功能有对应测试 | ❌ 无任何测试（见 P1-1） |
| 删除的测试有等价替代 | N/A |
| 边界 case 覆盖 | ❌ |
| 分布式改动有多卡测试 | N/A |

#### 代码质量与文档
| 检查项 | 结果 |
|--------|------|
| 无关改动分离 | ✅ |
| mega-PR 拆分 | ✅ 小 PR |
| Commit 原子性 | ✅ 单 commit |
| Commit message 准确性 | ✅ |
| PR description 说明动机和设计 | ❌ PR body 为空 |
| 日志频率控制 | ✅ warning 仅在初始化时 |

### 领域检查

#### A. 兼容性与配置
| 检查项 | 结果 |
|--------|------|
| proto 向后兼容 | ✅ 新增 optional 字段，field number 101 |
| 默认值影响 | ✅ 默认 nullopt，不影响现有行为 |
| 新增字段传播 | ❌ C++ Jsonize 缺失（见 P0-2） |
| pybind 默认值 | ✅ C++ 侧有默认值 |

#### B. 正确性与逻辑 — ❌ 见 P0-1
#### C. 线程安全与并发 — 全部 ✅
#### D. 性能 — 全部 ✅
#### E. 分布式 — 全部 ✅
#### F. 跨平台（ROCm/ARM） — 全部 ✅
#### G. 语言与框架特有 — 全部 ✅
#### H. 测试与 CI — ❌ 见 P1-1
#### I. 代码质量 — 全部 ✅

---

## Review 意见

### 问题

1. **BeamDedupLogitsProcessor 与 BaseLogitsProcessor 接口不兼容，无法编译** [P0]

   `BaseLogitsProcessor` 的接口使用 `rtp_llm::BufferPtr`，而 PR 中的 `BeamDedupLogitsProcessor` 使用 `torch::Tensor`，存在多处签名和类型不匹配，**代码无法通过编译**：

   **(a) 构造函数缺少 `DeviceBase*` 参数**
   基类 `BaseLogitsProcessor` 唯一构造函数签名为 `BaseLogitsProcessor(rtp_llm::DeviceBase* device)`（`BaseLogitsProcessor.h:13`），无默认构造函数。PR 中 `BeamDedupLogitsProcessor() = default` 无法调用基类构造函数，编译报错。

   **(b) `updateStatus` 签名不匹配**
   基类虚函数：`virtual void updateStatus(const rtp_llm::BufferPtr& new_tokens, int32_t num_new_tokens) = 0`
   PR 中 override：`void updateStatus(const torch::Tensor& new_tokens, int32_t num_new_tokens) override`
   参数类型 `BufferPtr` vs `torch::Tensor` 不同，不构成 override，编译报错。

   **(c) `process()` 内部将 `BufferPtr` 当作 `torch::Tensor` 使用**
   `SamplerInputs::logits` 类型为 `rtp_llm::BufferPtr`（`SampleInfos.h:34`），但 PR 代码中直接调用 `inputs.logits.size(0)`、`inputs.logits.size(1)`、`logits.max(/*dim=*/1, ...)`、`torch::ones(...)`、`logit_mask.scatter_(...)` 等 torch::Tensor API。`BufferPtr` 不支持这些操作。

   **(d) `maskLogits` 参数类型不匹配**
   基类 `maskLogits` 签名为 `void maskLogits(const rtp_llm::BufferPtr&, const rtp_llm::BufferPtr&)`，PR 中传入的是 `torch::Tensor`。

   建议：参考 `MultiSeqLogitsProcessor` 的实现模式：
   - 构造函数接受 `DeviceBase*` 并传递给基类
   - `fromGenerateInput` 接受 `DeviceBase*` 参数
   - 使用 `device_->allocateBuffer(...)` 分配 mask buffer
   - 使用 `BufferPtr` 的 `slice()`/`data<T>()` 操作数据
   - 调用基类 `maskLogits(BufferPtr, BufferPtr)`

2. **C++ GenerateConfig::Jsonize 缺少 beam_dedup_idx** [P0]

   `GenerateConfig.h` 新增了 `std::optional<int> beam_dedup_idx` 成员，但 `Jsonize()` 方法中没有添加 `JSONIZE_OPTIONAL(beam_dedup_idx)`。这意味着通过 JSON 接口（非 RPC）传入的请求无法设置此字段，字段永远为 `nullopt`。

   建议：在 `Jsonize()` 方法中其他 `JSONIZE_OPTIONAL` 调用附近添加：
   ```cpp
   JSONIZE_OPTIONAL(beam_dedup_idx);
   ```

3. **新功能无任何测试** [P1]

   `BeamDedupLogitsProcessor` 是全新的 logits processor，涉及 beam search 核心路径，但 PR 中没有任何单元测试或端到端测试。至少需要：
   - 单元测试：验证 `fromGenerateInput` 的各种边界条件（无 beam_dedup_idx、无 num_beams、idx 越界等）
   - 单元测试：验证 `process()` 在目标 step 正确 mask logits，非目标 step 不做操作
   - 端到端测试：验证 beam search + beam_dedup_idx 的完整推理流程

4. **Proto field number 从 54 跳到 101** [P2]

   `GenerateConfigPB` 最后一个字段是 `unique_key = 54`，新增字段直接用了 `101`。虽然 protobuf 允许不连续编号，但跳跃过大且无注释说明原因。如果是为了预留空间给其他字段，建议添加注释说明；否则建议使用 55。

5. **PR description 为空** [P2]

   PR body 为空，没有说明 beam dedup 的业务背景、设计思路和使用方式。建议补充：
   - beam_dedup_idx 的语义（在第 N 步 decode 时强制各 beam 选择各自的 argmax token）
   - 使用场景和预期效果
   - 与现有 beam search 逻辑的交互关系

### 小问题

- `BeamDedupLogitsProcessor.h` 析构函数建议用 `~BeamDedupLogitsProcessor() override = default;` 替代空实现 `{}`，与现代 C++ 风格一致
- `BeamDedupLogitsProcessor.cc` 使用 `make_shared` 后直接赋值私有成员，可考虑添加带参数的私有构造函数或 factory 内部构造，避免暴露内部状态

### 整体评价

PR 意图清晰——为 beam search 添加一个在指定 step 做 token 去重的 logits processor。但实现存在严重的编译错误：`BeamDedupLogitsProcessor` 的接口与 `BaseLogitsProcessor` 基类完全不兼容（构造函数、虚函数签名、数据类型均不匹配）。代码看起来是基于 torch::Tensor API 编写的，但 RTP-LLM 的 logits processor 框架使用的是 `rtp_llm::BufferPtr` 抽象。此外 Jsonize 序列化遗漏和零测试覆盖也是需要解决的问题。

❌ **存在阻塞/重要问题，不建议合入**

**Comment by LLLLKKKK** [comment]
## Code Review: PR #855 — feat:beam search biz decode

**审查者**: Claude Opus 4.6 | **日期**: 2026-04-14 | **变更**: +108/-0, 9 文件 | **CI**: 未找到运行记录

---

### P0 — 阻塞 (1)

**P0-1: `process()` 未使用 `start_idx`/`finish_idx`，batch 场景下会越界或处理错误数据**
> `BeamDedupLogitsProcessor.cc:27-43`

`process()` 直接使用 `inputs.logits.size(0)` 作为 batch_size，忽略了 `start_idx`/`finish_idx`。在 batched 推理中，`inputs.logits` 包含整个 batch 所有 stream 的 logits，当前 stream 只占 `[start_idx, finish_idx)` 范围。不使用这两个参数会导致对不属于当前 stream 的 logits 做错误 mask。

对比 `MultiSeqLogitsProcessor::process()` 的正确用法：
```cpp
size_t batch_size = finish_idx - start_idx;
auto logits = inputs.logits.narrow(0, start_idx, batch_size);
```

---

### P1 — 重要 (4)

**P1-1: `GenerateConfig::Jsonize()` 未序列化 `beam_dedup_idx`**
> `GenerateConfig.h` — `Jsonize()` 方法中缺少 `JSONIZE_OPTIONAL(beam_dedup_idx)`，JSON 路径请求无法设置该字段。

**P1-2: `debugString()` 未包含 `beam_dedup_idx`**
> 调试排查时无法看到该字段值。

**P1-3: 无任何测试覆盖**
> 新增 logits processor 包含 argmax + scatter mask 等非平凡 tensor 操作，需要单元测试覆盖正常路径、边界条件和 batch 切片。

**P1-4: CI 未运行**
> 未找到 branch `open_merge/855` 的 CI 记录，新增 C++ 代码需通过编译和基础测试。

---

### P2 — 建议 (3)

**P2-1**: Proto field number 从 56 跳到 101，建议使用连续编号或注释说明预留原因。
**P2-2**: `beam_dedup_idx == max_new_tokens - 1` 边界情况需确认与 EOS 处理逻辑的交互。
**P2-3**: PR body 为空，建议补充 beam dedup 的业务背景和使用示例。

---

### P3 — Nit (2)

**P3-1**: `current_step_` 在头文件和 `fromGenerateInput` 中重复初始化为 0。
**P3-2**: 建议用 private 构造函数替代 `fromGenerateInput` 中直接赋值 private 成员。

---

| 级别 | 数量 |
|------|------|
| P0 阻塞 | 1 |
| P1 重要 | 4 |
| P2 建议 | 3 |
| P3 Nit | 2 |

> P0 必须修复后才能合入。P1 强烈建议在本 PR 中一并解决。

<sub>Generated by RTP-LLM CI Bot (Claude Opus 4.6)</sub>

**Comment by LLLLKKKK** [comment]
🤖 AI Code Review — PR #855

## PR 概述

**Title:** `feat:beam search biz decode`
**Author:** zuozhen-ali
**规模:** 9 files, +108/-0

新增 `BeamDedupLogitsProcessor`，在 beam search 的指定 decode step 强制每个 beam 只保留 logits 最大的 token，实现 beam 去重。配置字段从 Python → proto → C++ 全链路打通。

---

## Review 意见

### 问题

1. **`process()` 忽略 `start_idx`/`finish_idx`，会污染同 batch 其他 stream 的 logits** [P0]

   `BeamDedupLogitsProcessor::process()` 直接使用 `inputs.logits.size(0)` 作为 batch_size，对整个 logits tensor 执行 argmax + mask 操作，完全忽略了 `start_idx` 和 `finish_idx` 参数。

   在 batched inference 中，多个 stream 的 logits 拼接在同一个 tensor 中。忽略这些参数会导致当前 processor 的 mask 操作覆盖同 batch 中其他 stream 的 logits。

   对比 `MultiSeqLogitsProcessor::process()` 的正确实现：
   ```cpp
   size_t batch_size = finish_idx - start_idx;
   auto logits = inputs.logits.narrow(0, start_idx, batch_size);
   ```

2. **C++ `GenerateConfig` 的 `Jsonize()` 和 pybind 未注册 `beam_dedup_idx`** [P1]

   `beam_dedup_idx` 已添加到 C++ struct 成员和 proto 定义中，但 `Jsonize()` 方法中缺少 `JSONIZE_OPTIONAL(beam_dedup_idx)`，pybind ConfigInit 中也未注册。导致通过 C++ HTTP API Server（JSON 路径）或 Python 直接调用 C++ engine（非 RPC 路径）时字段丢失。

3. **缺少单元测试** [P1]

   新增的 `BeamDedupLogitsProcessor` 没有任何测试。同目录下 `ThinkModeLogitsProcessor` 和 `TreeLogitsProcessor` 都有对应的 `*Test.cc`。建议至少覆盖 `fromGenerateInput()` 各校验分支、`process()` 在目标/非目标 step 的行为、`updateStatus()` 的 step 累加。

4. **`process()` 中 per-forward GPU 内存分配** [P2]

   直接在 GPU 上分配 mask tensor。虽然仅触发一次影响有限，但对比 `MultiSeqLogitsProcessor` 先在 CPU 构建再 `.to(kCUDA)` 的做法更规范。

5. **PR description 为空** [P2]

   建议补充功能动机、使用场景和设计决策说明。

### 小问题

- [P3] `generate_config.py` 中注释 `# beam dedup biz` 过于简略
- [P3] `BeamDedupLogitsProcessor.cc` 第 47 行有尾部空格

### 整体评价

PR 结构清晰，工厂模式扩展方式符合项目惯例。但存在 P0 级正确性问题（`process()` 忽略 batch 切片参数）、序列化路径不完整和无测试覆盖两个 P1 问题。建议修复后再合入。

**结论：存在阻塞/重要问题，不建议合入。**

**Comment by LLLLKKKK** [comment]
## AI Code Review — PR #855

**Summary:** P0/1 · P1/1 · P2/1 · P3/1

Review status: **review_blocking**

### Blocking Issues
- [P0] **process() ignores start_idx/finish_idx, masks logits for entire batch** @ `rtp_llm/cpp/models/logits_processor/BeamDedupLogitsProcessor.cc:27`
- [P1] **beam_dedup_idx not serialized in GenerateConfig::Jsonize()** @ `rtp_llm/cpp/engine_base/stream/GenerateConfig.h:89`

### Notes
- [P2] No unit test for BeamDedupLogitsProcessor @ `rtp_llm/cpp/models/logits_processor/BeamDedupLogitsProcessor.cc:1`
- [P3] Proto field number jumps from 54 to 101 @ `rtp_llm/cpp/model_rpc/proto/model_rpc_service.proto:99`

### Strengths
- Clean factory pattern with good early-return validation (negative idx, idx >= max_new_tokens, missing num_beams)
- Consistent use of existing TRANS_OPTIONAL macro and trans_option helper for proto/RPC plumbing
